### PR TITLE
Fix for JIRA issue APPM-1019

### PR DESCRIPTION
**Pull request by this**
Added "Create New Version" functionality for mobile apps.

### CODE REVIEW

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 1]
```diff
@@ -1,96 +1,143 @@
...
-      var shortName = context.params.type;
-      var fieldName = context.params.fieldName;
-      var fieldValue=context.params.fieldValue;
-    }
-	
-    try{
-
-        var contextVal = '/'+fieldValue;
-        var oldContextVal = '';
+include("/extensions/webapp/modules/jagg/jagg.jag");
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:10]: Why we depend on extensions jagg modules in API level file? We can have /jagg/jagg.jag module in publisher jaggery app root.

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 2]
```diff
@@ -1,96 +1,143 @@
...
+                    status: "Validation Error",
+                    validationResults: result.validationResults
+                };
+                response.content = stringify(content);
+                return;
+            }
+            response.status = 200;
+            response.content = stringify({status: "Success"});
+        }
+    });
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:73]: L24 to L42 and L47 to L72 codes are look like same. Cann't we reuse the code by using one function?

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 3]
```diff
@@ -1,96 +1,143 @@
...
+    routeManager.register('GET', 'publisher', '/publisher/api/validations/assets/{type}/{fieldName}/{fieldValue}', function (context) {
 
-    /**
-* Validate whether the context already exists
-*/
-function isContextExist(fieldValue,oldContext){
-    return apiProvider.isContextExist(fieldValue,oldContext);
-}
+        if (context.params) {
+            var shortName = context.params.type;
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:95]: Please rename shortName variable name

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 4]
```diff
@@ -1,96 +1,143 @@
...
 
+        try {
+            var contextVal = '/' + fieldValue;
+            var oldContextVal = '';
+            var isContextExists = isContextExist(contextVal, oldContextVal);
+            print(isContextExists);
 
+        } catch (e) {
+            log.error('An asset of type: ' + shortName
+                      + ' could not be created.The following exception was thrown');
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:108]: created.The -> created. The

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 5]
```diff
@@ -1,96 +1,143 @@
...
+            log.error(e);
+            response.sendError(500, 'An asset of type: ' + shortName
+                                    + ' could not be created.Please check the server logs.');
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:111]: "Please check the server logs." The end user will not be able to view server logs.

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 6]
```diff
@@ -1,96 +1,143 @@
...
 
-var securityModule=require('/modules/security/security.manager.js').securityManagementModule();
+    });
 
-var sm=securityModule.cached();
+    /**
+     * Validate whether the context already exists
+     */
+    function isContextExist(fieldValue, oldContext) {
+        return apiProvider.isContextExist(fieldValue, oldContext);
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:120]: What happen if we got an exception when calling the host object?

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 7]
```diff
@@ -1,96 +1,143 @@
...
 
-//This will short circuit the handling of the requests
-var passed=sm.check(session,function(){
-    response.sendError(401);
-});
 
-//Stop servicing the request if the check failed
-if(passed){
+    var securityModule = require('/modules/security/security.manager.js').securityManagementModule();
+    var sm = securityModule.cached();
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:125]: sm? variable name

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/themes/mobileapp/helpers/edit-mobileapp.js [hunk 8]
```diff
@@ -1,41 +1,38 @@
...
+    };
 };
 
-var selectCategory = function(data) {
-	var selected, 
-		arr=[],
-		currentCategory = data.artifact.attributes['overview_category'],
-		categories = selectCategories(data.data.fields);
+var selectCategory = function (data) {
+    var arr = [];
```
Comment (sajithab) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/themes/mobileapp/helpers/edit-mobileapp.js:12]: arr? variable name.

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 1 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:10]: This is merely a code formatting diff. '/jagg/jagg.jag' module was already included in [old line 10](https://github.com/wso2/carbon-appmgt/pull/96/files#diff-a83986c9004a1801852de58518c36a07L10).

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 3 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:95]: fixed in commit c557bd157b4574696302bd32f5020aac06c3b91c of PR #241

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 4 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:108]: fixed in commit 46d619452d3d5b4396b7afff8d0f65db7777111c of PR #241

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 5 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:111]: fixed in commit 46d619452d3d5b4396b7afff8d0f65db7777111c of PR #241

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag [hunk 9]
```diff
@@ -1,96 +1,143 @@
...
-
             var oldContextVal = '';
             var isContextExists = isContextExist(contextVal, oldContextVal);
             print(isContextExists);
-
-        }
-        catch (e) {
-            log.debug('An asset of type: ' + shortName + ' could not be created.The following exception was thrown: ' + e);
-            response.sendError(500, 'An asset of type: ' + shortName + ' could not be created.Please check the server logs.');
+        } catch (e) {
```
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:82]: fixed error message in commit b613c0c77f021a82c6ca7d0505b64ce41013e6f5 of PR #241

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 6 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:120]: `isContextExist` function should be wrapped in a try-catch block when calling. See line [80](https://github.com/wso2/carbon-appmgt/pull/96/files#diff-a83986c9004a1801852de58518c36a07R80) & [103](https://github.com/wso2/carbon-appmgt/pull/96/files#diff-a83986c9004a1801852de58518c36a07R103).

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag (hunk 7 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/apis/v1/validations_api_router.jag:125]: This line is not a newly added one. It was there & merely formatted (see [old line 76](https://github.com/wso2/carbon-appmgt/pull/96/files#diff-a83986c9004a1801852de58518c36a07L76)).
In other '*_api_router.jag' files, `sm` variable name is used for security manager object. Same pattern followed here.

File: features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/themes/mobileapp/helpers/edit-mobileapp.js (hunk 8 above)
Comment (this) [inline features/appmgt/org.wso2.carbon.appmgt.publisher.feature/src/main/resources/publisher/themes/mobileapp/helpers/edit-mobileapp.js:12]: fixed in commit bb1bdf04004cbfa87bb91eda8c146e49e6da7516 of PR #241

**Pull request merged**